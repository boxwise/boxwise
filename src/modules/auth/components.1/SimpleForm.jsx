import React, {useState, useEffect } from "react";

export const useForm = (callback, initialValues) => {
  const [values, setValues] = useState(initialValues || {});
  const [isSubmitting, setIsSubmitting] = useState(false);

  const handleSubmit = event => {
    if (event) event.preventDefault();
    setIsSubmitting(true);
    Promise.resolve(callback(values)).
      then(setIsSubmitting(false));
  };

  const handleChange = event => {
    event.persist();
    setValues(values => ({
      ...values,
      [event.target.name]: event.target.value
    }));
  };

  const attachToForm = (name) => {
    return {
      onChange: handleChange,
      value: values[name] || "",
    };
  };

  useEffect(() => {
    const resetFormWhenDefaultValuesChange = () => {
      setValues(initialValues || {});
      setIsSubmitting(false);
    };
    resetFormWhenDefaultValuesChange();
  }, [initialValues]);

  return {
    attachToForm,
    handleSubmit,
    values,
    isSubmitting
  };
};

const SignInForm = ({ userSignIn }) => {
  const { attachToForm, handleSubmit } = useForm(userSignIn);

  return (
    <form onSubmit={handleSubmit}>
      <input 
        type="text"
        name="password2"
        data-testid="password2"
        {...attachToForm("password2")}
      />
      <input 
        type="text"
        name="email"
        data-testid="email"
        {...attachToForm("email")}
      />
      <input type="submit" data-testid="signInButton" value="Sign in" />
    </form>
  );
};

export default SignInForm;
