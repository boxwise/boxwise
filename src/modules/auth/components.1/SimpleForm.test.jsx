
import React from "react";
import { render, fireEvent, cleanup, wait, act } from "react-testing-library";

import SimpleForm from "./SimpleForm";

beforeEach(cleanup);

describe("SignInForm", () => {
  let container;
  let getByTestId;

  const handler = jest.fn(() => Promise.resolve());

  beforeEach(() => {
    act(() => {
      ({ container, getByTestId } = render(<SimpleForm userSignIn={handler} />));
    });
  });

  // the first test always fails.

  it("with only email", async () => {
    fireEvent.change(getByTestId("email"), { target: { value: "testX" } });

    fireEvent.click(getByTestId("signInButton"));

    expect(getByTestId("email").value).toEqual("testX");
  });

  it("with only password", async () => {
    fireEvent.change(getByTestId("password2"), { target: { value: "passw0rd" } });

    fireEvent.click(getByTestId("signInButton"));

    expect(getByTestId("password2").value).toEqual("passw0rd");
  });
});
