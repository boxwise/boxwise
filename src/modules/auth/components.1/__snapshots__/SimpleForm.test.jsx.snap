// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm email and password: after submit 1`] = `
<div>
  <form>
    <div
      class="MuiFormControl-root-185 MuiFormControl-marginNormal-186 MuiFormControl-fullWidth-188"
    >
      <label
        class="MuiFormLabel-root-200 MuiFormLabel-filled-204 MuiInputLabel-root-189 MuiInputLabel-formControl-194 MuiInputLabel-animated-197 MuiInputLabel-shrink-196"
        data-shrink="true"
        for="password2"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root-220 MuiInput-root-207 MuiInput-underline-211 MuiInputBase-fullWidth-229 MuiInput-fullWidth-214 MuiInputBase-formControl-221 MuiInput-formControl-208"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-230 MuiInput-input-215 MuiInputBase-inputType-233 MuiInput-inputType-218"
          data-testid="password2"
          id="password2"
          name="password2"
          type="email"
          value="passw0rd"
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-185 MuiFormControl-marginNormal-186 MuiFormControl-fullWidth-188"
    >
      <label
        class="MuiFormLabel-root-200 MuiFormLabel-filled-204 MuiInputLabel-root-189 MuiInputLabel-formControl-194 MuiInputLabel-animated-197 MuiInputLabel-shrink-196"
        data-shrink="true"
        for="email"
      >
        Email address
      </label>
      <div
        class="MuiInputBase-root-220 MuiInput-root-207 MuiInput-underline-211 MuiInputBase-fullWidth-229 MuiInput-fullWidth-214 MuiInputBase-formControl-221 MuiInput-formControl-208"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-230 MuiInput-input-215 MuiInputBase-inputType-233 MuiInput-inputType-218"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value="testX"
        />
      </div>
    </div>
    <div
      class="SubmitButton-root-237 SubmitButton-submitInPage-240"
    >
      <button
        class="MuiButtonBase-root-267 MuiButton-root-241 MuiButton-contained-252 MuiButton-containedPrimary-253 MuiButton-raised-255 MuiButton-raisedPrimary-256"
        data-testid="signInButton"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-242"
        >
          Sign In
        </span>
        <span
          class="MuiTouchRipple-root-270"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`SignInForm email and password: before submit 1`] = `
<div>
  <form>
    <div
      class="MuiFormControl-root-185 MuiFormControl-marginNormal-186 MuiFormControl-fullWidth-188"
    >
      <label
        class="MuiFormLabel-root-200 MuiFormLabel-filled-204 MuiInputLabel-root-189 MuiInputLabel-formControl-194 MuiInputLabel-animated-197 MuiInputLabel-shrink-196"
        data-shrink="true"
        for="password2"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root-220 MuiInput-root-207 MuiInput-underline-211 MuiInputBase-fullWidth-229 MuiInput-fullWidth-214 MuiInputBase-formControl-221 MuiInput-formControl-208"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-230 MuiInput-input-215 MuiInputBase-inputType-233 MuiInput-inputType-218"
          data-testid="password2"
          id="password2"
          name="password2"
          type="email"
          value="passw0rd"
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-185 MuiFormControl-marginNormal-186 MuiFormControl-fullWidth-188"
    >
      <label
        class="MuiFormLabel-root-200 MuiFormLabel-filled-204 MuiInputLabel-root-189 MuiInputLabel-formControl-194 MuiInputLabel-animated-197 MuiInputLabel-shrink-196"
        data-shrink="true"
        for="email"
      >
        Email address
      </label>
      <div
        class="MuiInputBase-root-220 MuiInput-root-207 MuiInput-underline-211 MuiInputBase-fullWidth-229 MuiInput-fullWidth-214 MuiInputBase-formControl-221 MuiInput-formControl-208"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-230 MuiInput-input-215 MuiInputBase-inputType-233 MuiInput-inputType-218"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value="testX"
        />
      </div>
    </div>
    <div
      class="SubmitButton-root-237 SubmitButton-submitInPage-240"
    >
      <button
        class="MuiButtonBase-root-267 MuiButton-root-241 MuiButton-contained-252 MuiButton-containedPrimary-253 MuiButton-raised-255 MuiButton-raisedPrimary-256"
        data-testid="signInButton"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-242"
        >
          Sign In
        </span>
        <span
          class="MuiTouchRipple-root-270"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`SignInForm with only email: after submit 1`] = `
<div>
  <form>
    <div
      class="MuiFormControl-root-1 MuiFormControl-marginNormal-2 MuiFormControl-fullWidth-4"
    >
      <label
        class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink="false"
        for="password2"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputType-49 MuiInput-inputType-34"
          data-testid="password2"
          id="password2"
          name="password2"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-1 MuiFormControl-marginNormal-2 MuiFormControl-fullWidth-4"
    >
      <label
        class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink="false"
        for="email"
      >
        Email address
      </label>
      <div
        class="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputType-49 MuiInput-inputType-34"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="SubmitButton-root-53 SubmitButton-submitInPage-56"
    >
      <button
        class="MuiButtonBase-root-83 MuiButton-root-57 MuiButton-contained-68 MuiButton-containedPrimary-69 MuiButton-raised-71 MuiButton-raisedPrimary-72"
        data-testid="signInButton"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-58"
        >
          Sign In
        </span>
        <span
          class="MuiTouchRipple-root-86"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`SignInForm with only email: before submit 1`] = `
<div>
  <form>
    <div
      class="MuiFormControl-root-1 MuiFormControl-marginNormal-2 MuiFormControl-fullWidth-4"
    >
      <label
        class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink="false"
        for="password2"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputType-49 MuiInput-inputType-34"
          data-testid="password2"
          id="password2"
          name="password2"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-1 MuiFormControl-marginNormal-2 MuiFormControl-fullWidth-4"
    >
      <label
        class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink="false"
        for="email"
      >
        Email address
      </label>
      <div
        class="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputType-49 MuiInput-inputType-34"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="SubmitButton-root-53 SubmitButton-submitInPage-56"
    >
      <button
        class="MuiButtonBase-root-83 MuiButton-root-57 MuiButton-contained-68 MuiButton-containedPrimary-69 MuiButton-raised-71 MuiButton-raisedPrimary-72"
        data-testid="signInButton"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-58"
        >
          Sign In
        </span>
        <span
          class="MuiTouchRipple-root-86"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`SignInForm with only password: after submit 1`] = `
<div>
  <form>
    <div
      class="MuiFormControl-root-93 MuiFormControl-marginNormal-94 MuiFormControl-fullWidth-96"
    >
      <label
        class="MuiFormLabel-root-108 MuiFormLabel-filled-112 MuiInputLabel-root-97 MuiInputLabel-formControl-102 MuiInputLabel-animated-105 MuiInputLabel-shrink-104"
        data-shrink="true"
        for="password2"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root-128 MuiInput-root-115 MuiInput-underline-119 MuiInputBase-fullWidth-137 MuiInput-fullWidth-122 MuiInputBase-formControl-129 MuiInput-formControl-116"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-138 MuiInput-input-123 MuiInputBase-inputType-141 MuiInput-inputType-126"
          data-testid="password2"
          id="password2"
          name="password2"
          type="email"
          value="passw0rd"
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-93 MuiFormControl-marginNormal-94 MuiFormControl-fullWidth-96"
    >
      <label
        class="MuiFormLabel-root-108 MuiInputLabel-root-97 MuiInputLabel-formControl-102 MuiInputLabel-animated-105"
        data-shrink="false"
        for="email"
      >
        Email address
      </label>
      <div
        class="MuiInputBase-root-128 MuiInput-root-115 MuiInput-underline-119 MuiInputBase-fullWidth-137 MuiInput-fullWidth-122 MuiInputBase-formControl-129 MuiInput-formControl-116"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-138 MuiInput-input-123 MuiInputBase-inputType-141 MuiInput-inputType-126"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="SubmitButton-root-145 SubmitButton-submitInPage-148"
    >
      <button
        class="MuiButtonBase-root-175 MuiButton-root-149 MuiButton-contained-160 MuiButton-containedPrimary-161 MuiButton-raised-163 MuiButton-raisedPrimary-164"
        data-testid="signInButton"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-150"
        >
          Sign In
        </span>
        <span
          class="MuiTouchRipple-root-178"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`SignInForm with only password: before submit 1`] = `
<div>
  <form>
    <div
      class="MuiFormControl-root-93 MuiFormControl-marginNormal-94 MuiFormControl-fullWidth-96"
    >
      <label
        class="MuiFormLabel-root-108 MuiFormLabel-filled-112 MuiInputLabel-root-97 MuiInputLabel-formControl-102 MuiInputLabel-animated-105 MuiInputLabel-shrink-104"
        data-shrink="true"
        for="password2"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root-128 MuiInput-root-115 MuiInput-underline-119 MuiInputBase-fullWidth-137 MuiInput-fullWidth-122 MuiInputBase-formControl-129 MuiInput-formControl-116"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-138 MuiInput-input-123 MuiInputBase-inputType-141 MuiInput-inputType-126"
          data-testid="password2"
          id="password2"
          name="password2"
          type="email"
          value="passw0rd"
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-93 MuiFormControl-marginNormal-94 MuiFormControl-fullWidth-96"
    >
      <label
        class="MuiFormLabel-root-108 MuiInputLabel-root-97 MuiInputLabel-formControl-102 MuiInputLabel-animated-105"
        data-shrink="false"
        for="email"
      >
        Email address
      </label>
      <div
        class="MuiInputBase-root-128 MuiInput-root-115 MuiInput-underline-119 MuiInputBase-fullWidth-137 MuiInput-fullWidth-122 MuiInputBase-formControl-129 MuiInput-formControl-116"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input-138 MuiInput-input-123 MuiInputBase-inputType-141 MuiInput-inputType-126"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="SubmitButton-root-145 SubmitButton-submitInPage-148"
    >
      <button
        class="MuiButtonBase-root-175 MuiButton-root-149 MuiButton-contained-160 MuiButton-containedPrimary-161 MuiButton-raised-163 MuiButton-raisedPrimary-164"
        data-testid="signInButton"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-150"
        >
          Sign In
        </span>
        <span
          class="MuiTouchRipple-root-178"
        />
      </button>
    </div>
  </form>
</div>
`;
